pkg/team/api/api.go

```diff
@@ -0,0 +1,484 @@
+// limitations under the License.
+
+package api
+
+import (
+	"context"
+	"errors"
+	"strconv"
+	"strings"
+
+	"go.uber.org/zap"
+	"google.golang.org/genproto/googleapis/rpc/errdetails"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	gstatus "google.golang.org/grpc/status"
+
+	accclient "github.com/bucketeer-io/bucketeer/pkg/account/client"
+	domainevent "github.com/bucketeer-io/bucketeer/pkg/domainevent/domain"
+	"github.com/bucketeer-io/bucketeer/pkg/locale"
+	"github.com/bucketeer-io/bucketeer/pkg/log"
+	"github.com/bucketeer-io/bucketeer/pkg/pubsub/publisher"
+	"github.com/bucketeer-io/bucketeer/pkg/role"
+	"github.com/bucketeer-io/bucketeer/pkg/storage/v2/mysql"
+	"github.com/bucketeer-io/bucketeer/pkg/team/domain"
+	"github.com/bucketeer-io/bucketeer/pkg/team/storage"
+	accountproto "github.com/bucketeer-io/bucketeer/proto/account"
+	accproto "github.com/bucketeer-io/bucketeer/proto/account"
+	eventproto "github.com/bucketeer-io/bucketeer/proto/event/domain"
+	proto "github.com/bucketeer-io/bucketeer/proto/team"
+)
+
+var (
+	statusInternal         = gstatus.New(codes.Internal, "team: internal")
+	statusNameRequired     = gstatus.New(codes.InvalidArgument, "team: name must be specified")
+	statusInvalidCursor    = gstatus.New(codes.InvalidArgument, "team: cursor is invalid")
+	statusInvalidOrderBy   = gstatus.New(codes.InvalidArgument, "team: order_by is invalid")
+	statusUnauthenticated  = gstatus.New(codes.Unauthenticated, "team: unauthenticated")
+	statusPermissionDenied = gstatus.New(codes.PermissionDenied, "team: permission denied")
+	statusTeamNotFound     = gstatus.New(codes.NotFound, "team: not found")
+)
+
+type options struct {
+	logger *zap.Logger
+}
+
+type Option func(*options)
+
+func WithLogger(l *zap.Logger) Option {
+	return func(opts *options) {
+		opts.logger = l
+	}
+}
+
+type TeamService struct {
+	mysqlClient   mysql.Client
+	teamStorage   storage.TeamStorage
+	accountClient accclient.Client
+	publisher     publisher.Publisher
+	opts          *options
+	logger        *zap.Logger
+}
+
+func NewTeamService(
+	mysqlClient mysql.Client,
+	accountClient accclient.Client,
+	publisher publisher.Publisher,
+	opts ...Option,
+) *TeamService {
+	dopts := &options{
+		logger: zap.NewNop(),
+	}
+	for _, opt := range opts {
+		opt(dopts)
+	}
+	return &TeamService{
+		mysqlClient:   mysqlClient,
+		teamStorage:   storage.NewTeamStorage(mysqlClient),
+		accountClient: accountClient,
+		publisher:     publisher,
+		opts:          dopts,
+		logger:        dopts.logger.Named("api"),
+	}
+}
+
+func (s *TeamService) Register(server *grpc.Server) {
+	proto.RegisterTeamServiceServer(server, s)
+}
+
+func (s *TeamService) CreateTeam(
+	ctx context.Context,
+	req *proto.CreateTeamRequest,
+) (*proto.CreateTeamResponse, error) {
+	localizer := locale.NewLocalizer(ctx)
+	editor, err := s.checkOrganizationRole(
+		ctx, req.OrganizationId,
+		accproto.AccountV2_Role_Organization_MEMBER, localizer)
```

Comment:
Since members can only be created by an admin, we should use the same rule for teams.
Same for the update and delete APIs.